Add a pyro.contrib.cevae module

fritzo opened the pull request:
This refactors examples/cevae.py into a new module `pyro.contrib.cevae`. The top level interface is very simple and requires no Pyro:
```py
cevae = CEVAE(feature_dim=5)
cevae.fit(x_train, t_train, y_train)
ite = cevae.ite(x_test)  # individual treatment effect
ate = ite.mean()         # average treatment effect
```
Lower level components `Model`, `Guide`, `TraceCausalEffect_ELBO` are exposed for advanced users to customize. The `CEVAE` module can be jit traced for serving the `.ite()` method as follows:
```py
traced_cevae = cevae.jit_trace()
# ...traced_cevae can be saved and loaded...
ite = traced_cevae.ite(x_test)
```

## Reviewers

@eb8680 can you please review general interface and examples?

@neerajprad can you please review jit tracing logic?

See sphinx docs [here](http://fritzo.org/temp/pyro/html/contrib.cevae.html).

## Tested
- [x] refactoring is exercised by `pytest tests/test_examples.py -k cevae`
- [x] added a smoke test
- [x] added serialization tests
- [x] tested on a large real-world dataset (1.2M x 112)

eb8680 commented on examples/cevae.py:
```diff
@@ -1,11 +1,7 @@
 """
```
Move to `examples/contrib/cevae/`

fritzo replied:
done.

fritzo commented on pyro/contrib/cevae/__init__.py:
```diff
@@ -0,0 +1,396 @@
+                y1 = poutine.replay(self.model, tr.trace)(x)
+        return (y1 - y0).mean(0)
+
+    def jit_trace(self):
```
@neerajprad does this look like a reasonable interface?

neerajprad replied:
This looks good. As discussed, how about naming this `to_script_module()` or `get_script_module()`?

fritzo replied:
done.

fritzo commented:
@eb8680 ready for review again. Main changes since last time are:
- numerical stability fixes for practical application
- include an `outcome_dist` kwarg and a few different distributions; I found `student` and `laplace` seemed to work better than `normal`. The dispatch mechanism is simple and user-extensible, relying on `.__subclasses__()`.
- more tests for serialization

fritzo commented on pyro/contrib/cevae/__init__.py:
```diff
@@ -0,0 +1,591 @@
+
+        :param ~torch.Tensor x:
+        :param ~torch.Tensor t:
+        :param ~to rch.Tensor y:
```
done
